### PR TITLE
upgrading to python 3.8

### PR DESCRIPTION
**Pull request by Abhinavfreecodecamp**
**What this PR does / why we need it**:

This PR is to upgrade pre-packaged servers from python 3.7 to python 3.8.

**Which issue(s) this PR fixes**:

Issue can be found [here](https://github.com/SeldonIO/seldon-core/issues/4803) 

**Special notes for your reviewer**:

@adriangonz Let's work on this PR to upgrade all pre-packaged servers.

### CODE REVIEW

**Comment by adriangonz** [comment]
Hey @Abhinavfreecodecamp ,

Keep in mind that the explainer / detect servers will be upgraded on a separate PR. If you're happy to make the changes to the pre-packaged servers here, I can review once they're done - there's no need to collaborate on the same PR as that tends to confuse things.

I'd also flag this as a WIP PR until it's ready.

**Comment by Abhinavfreecodecamp** [comment]
Hi @adriangonz, 

Sure, let me remove any changes to done to explainer/detect servers in this PR and this PR changes will be limited to` /server` dir.  Yes WIP for this PR is fine until its ready.

**Comment by adriangonz** [comment]
Hey @Abhinavfreecodecamp ,

Following up on this one, would you be able to continue working on this PR? If not, that's totally understandable - we can always pick this up internally to finalise it ahead of the next release of SC.

**Comment by Abhinavfreecodecamp** [comment]
@adriangonz , This PR will be completed by today. I was testing it locally.

File: servers/sklearnserver/sklearnserver/requirements.txt [hunk 1]
```diff
@@ -1,4 +1,4 @@
 seldon_core
-scikit-learn == 0.24.2
-numpy >= 1.8.2
+scikit-learn == 0.23.0
```
Comment (adriangonz) [inline servers/sklearnserver/sklearnserver/requirements.txt:2]: Do you know why the downgrade of both `scikit-learn` and `numpy` was required?

File: servers/xgboostserver/xgboostserver/requirements.txt [hunk 2]
```diff
@@ -1,3 +1,3 @@
-scikit-learn == 0.20.3
-numpy >= 1.8.2
+scikit-learn == 0.23.0
```
Comment (adriangonz) [inline servers/xgboostserver/xgboostserver/requirements.txt:1]: Similar as above - do you know why the downgrade was required?

**Review by adriangonz — COMMENTED** [review]
Nice one @Abhinavfreecodecamp ! Thanks for taking the time for this contribution.

I've added a couple comments below. Other than those, I think we're good to start kicking off the CI pipelines (which will build the images, run integration tests, etc. ).

File: servers/sklearnserver/sklearnserver/requirements.txt (hunk 1 above)
Comment (Abhinavfreecodecamp) [reply servers/sklearnserver/sklearnserver/requirements.txt]: @adriangonz sorry about this wrong push. I have fixed the sklearn, numpy, xgb versions.

File: servers/xgboostserver/xgboostserver/requirements.txt (hunk 2 above)
Comment (Abhinavfreecodecamp) [reply servers/xgboostserver/xgboostserver/requirements.txt]: @adriangonz same sorry about this wrong push. I have fixed the sklearn, numpy, xgb versions.

File: servers/sklearnserver/sklearnserver/requirements.txt [hunk 3]
```diff
@@ -1,4 +1,4 @@
 seldon_core
-scikit-learn == 0.24.2
-numpy >= 1.8.2
+scikit-learn == 1.1.0
```
Comment (adriangonz) [inline servers/sklearnserver/sklearnserver/requirements.txt:2]: Would it be possible to keep the previous versions?

**Review by adriangonz — CHANGES_REQUESTED** [review]

**Comment by seldondev** [comment]
@Abhinavfreecodecamp: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 1a199da5af558d4701a39409a32e2a1495814e83 | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-4910-integration-jnwpv) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@Abhinavfreecodecamp: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | 1a199da5af558d4701a39409a32e2a1495814e83 | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-4910-integration-jnwpv) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: servers/sklearnserver/sklearnserver/requirements.txt (hunk 3 above)
Comment (Abhinavfreecodecamp) [reply servers/sklearnserver/sklearnserver/requirements.txt:2]: Yes it will be possible.

**Comment by seldondev** [comment]
@Abhinavfreecodecamp: Cannot trigger testing until a trusted user reviews the PR and leaves an `/ok-to-test` message.

<details>

In response to [this](https://github.com/SeldonIO/seldon-core/pull/4910#issuecomment-1614930914):

>/test integration


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository.
</details>

File: servers/sklearnserver/sklearnserver/requirements.txt (hunk 3 above)
Comment (adriangonz) [reply servers/sklearnserver/sklearnserver/requirements.txt:2]: I would keep it to the same versions as in `master` to reduce the number of changes

Comment (adriangonz) [reply servers/sklearnserver/sklearnserver/requirements.txt:2]: i.e.

`scikit-learn==0.24.2`, `numpy>=1.8.2` and same would apply to the xgboost reqs

Comment (Abhinavfreecodecamp) [reply servers/sklearnserver/sklearnserver/requirements.txt:2]: @adriangonz sure let me revert back changes.

Comment (Abhinavfreecodecamp) [reply servers/sklearnserver/sklearnserver/requirements.txt:2]: @adriangonz in master for xgboost server the requirements are follows should I keep it same ?
```
scikit-learn == 0.20.3
numpy >= 1.8.2
xgboost == 1.4.2
```

Comment (adriangonz) [reply servers/sklearnserver/sklearnserver/requirements.txt:2]: Yeah, let's keep the same as in `master` for both

Comment (Abhinavfreecodecamp) [reply servers/sklearnserver/sklearnserver/requirements.txt]: @adriangonz Done.

**Review by adriangonz — APPROVED** [review]
Thanks for making those changes! PR looks good - once tests are green, it should be good to go.

**Comment by seldondev** [comment]
@Abhinavfreecodecamp: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | eb3b3fbb95385756079e951db1a0006d801c02d9 | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-4910-integration-r7dqw) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@Abhinavfreecodecamp: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | eb3b3fbb95385756079e951db1a0006d801c02d9 | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-4910-integration-k4lm6) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by Abhinavfreecodecamp** [comment]
@adriangonz any idea why integration has failed as for the mentioned failed commit, it just revert back the changes ?

**Comment by Abhinavfreecodecamp** [comment]
@adriangonz can you please help here, so that we can merge this PR.

**Comment by adriangonz** [comment]
@Abhinavfreecodecamp most likely is due to flakiness on the integration tests setup

**Comment by seldondev** [comment]
@Abhinavfreecodecamp: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | eb3b3fbb95385756079e951db1a0006d801c02d9 | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-4910-integration-l9lrh) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by Abhinavfreecodecamp** [comment]
@adriangonz I think it failed again 😅 , is it possible to share logs for the test so I can help with that.

**Comment by adriangonz** [comment]
@Abhinavfreecodecamp so far it's just failing before the setup stage - so nothing to do with the scope of this PR. Tests seem flakier than usual for some reason.

Once it goes beyond that I'll share the logs if failures are relevant to this PR :+1:

**Comment by seldondev** [comment]
@Abhinavfreecodecamp: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | eb3b3fbb95385756079e951db1a0006d801c02d9 | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-4910-integration-qc62w) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by Abhinavfreecodecamp** [comment]
@adriangonz it failed again 😅. ?

**Comment by adriangonz** [comment]
Ok - now it's a more relevant error:

```
___________________________ TestPrepack.test_xgboost ___________________________
[gw0] linux -- Python 3.7.10 /opt/conda/bin/python

self = <test_prepackaged_servers.TestPrepack object at 0x7f42690ece10>
namespace = 'test-xgboost'

    def test_xgboost(self, namespace):
        spec = "../../servers/xgboostserver/samples/iris.yaml"
        retry_run(f"kubectl apply -f {spec}  -n {namespace}")
        wait_for_status("xgboost", namespace)
>       wait_for_rollout("xgboost", namespace)

test_prepackaged_servers.py:108: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
seldon_e2e_utils.py:161: in wait_for_rollout
    wait_for_deployment(deployment_name, namespace, attempts, sleep)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

deployment_name = 'xgboost-default-0-classifier', namespace = 'test-xgboost'
attempts = 50, sleep = 5

    def wait_for_deployment(deployment_name, namespace, attempts=50, sleep=5):
        logging.info(f"Waiting for deployment {deployment_name}")
        for _ in range(attempts):
            ret = run(
                f"kubectl rollout status -n {namespace} deploy/{deployment_name}",
                shell=True,
            )
            if ret.returncode == 0:
                logging.info(f"Successfully waited for deployment {deployment_name}")
                break
            logging.warning(f"Unsuccessful wait command but retrying for {deployment_name}")
            time.sleep(sleep)
        assert (
            ret.returncode == 0
>       ), f"Wait for rollout of {deployment_name} failed: non-zero return code"
E       AssertionError: Wait for rollout of xgboost-default-0-classifier failed: non-zero return code

seldon_e2e_utils.py:140: AssertionError
```

**Comment by Abhinavfreecodecamp** [comment]
@adriangonz let me setup the tests locally.

**Comment by adriangonz** [comment]
Sure @Abhinavfreecodecamp  - to run the tests locally, you will need to run the `make -C testing/scripts kind_create_cluster kind_setup` rules. These assume that KinD is already present locally (you will need to install it otherwise).

Besides that, you will also need to build and load your new images onto the cluster (which you can do with `make -C testing/scripts kind_build_prepackaged`).

Once the above is done (i.e. SC is running on a Kind cluster, where your new images have been loaded) you will be able to run the failing test.

**Comment by Abhinavfreecodecamp** [comment]
Hi @adriangonz ,Please see result below I ran for xgboost I did need to comment rest of the test cases as my docker memory ran out. For xgboost result went fine take a look below. (skipped warning logs)
`
~/Documents/seldon-core/testing/scripts on  upgrading-pre-pacakged-server! ⌚ 22:01:08
$ pytest test_prepackaged_servers.py
================================================================================ test session starts ================================================================================
platform darwin -- Python 3.8.16, pytest-7.3.1, pluggy-1.0.0
rootdir: /Users/abhinav.singh/Documents/seldon-core/testing/scripts
configfile: setup.cfg
plugins: anyio-3.7.0
collected 1 item                                                                                                                                                                    

test_prepackaged_servers.py .                                                                                                                                                 [100%]
==================================================================== 1 passed, 100 warnings in 65.13s (0:01:05) =====================================================================
`

also below is output while running `../../servers/xgboostserver/samples/iris.yaml` directly. In both scenerios it is running fine. Hence I am not able to replicate this error, is it possible for you to try at your end ?

Screenshot 2023-07-04 at 10 10 04 PM

**Comment by Abhinavfreecodecamp** [comment]
Hi @adriangonz is it possible this is happening due to k8s configuration and not due to changes in this PR ?

**Comment by Abhinavfreecodecamp** [comment]
hi @adriangonz sorry to ask you again, but if it is possible could you test this PR on your local ?

**Comment by adriangonz** [comment]
Hey @Abhinavfreecodecamp ,

Got the message the first time :slightly_smiling_face: 

We'll try it out once we get a chance.

**Comment by Abhinavfreecodecamp** [comment]
> Hey @Abhinavfreecodecamp ,
> 
> Got the message the first time :slightly_smiling_face: 
> 
> We'll try it out once we get a chance.

Thanks @adriangonz

**Comment by adriangonz** [comment]
Hey @Abhinavfreecodecamp ,

Just had a look and was able to trigger the following error when running the XGBoost test:

```
Traceback (most recent call last):
  File "/opt/conda/bin/seldon-core-microservice", line 8, in <module>
    sys.exit(main())
  File "/opt/conda/lib/python3.8/site-packages/seldon_core/microservice.py", line 610, in main
    interface_file = importlib.import_module(args.interface_name)
  File "/opt/conda/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 848, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/microservice/XGBoostServer.py", line 8, in <module>
    import xgboost as xgb
  File "/opt/conda/lib/python3.8/site-packages/xgboost/__init__.py", line 9, in <module>
    from .core import DMatrix, DeviceQuantileDMatrix, Booster
  File "/opt/conda/lib/python3.8/site-packages/xgboost/core.py", line 23, in <module>
    from .compat import (STRING_TYPES, DataFrame, py_str, PANDAS_INSTALLED,
  File "/opt/conda/lib/python3.8/site-packages/xgboost/compat.py", line 46, in <module>
    from sklearn.base import BaseEstimator
  File "/opt/conda/lib/python3.8/site-packages/sklearn/__init__.py", line 64, in <module>
    from .base import clone
  File "/opt/conda/lib/python3.8/site-packages/sklearn/base.py", line 14, in <module>
    from .utils.fixes import signature
  File "/opt/conda/lib/python3.8/site-packages/sklearn/utils/__init__.py", line 14, in <module>
    from . import _joblib
  File "/opt/conda/lib/python3.8/site-packages/sklearn/utils/_joblib.py", line 22, in <module>
    from ..externals import joblib
  File "/opt/conda/lib/python3.8/site-packages/sklearn/externals/joblib/__init__.py", line 119, in <module>
    from .parallel import Parallel
  File "/opt/conda/lib/python3.8/site-packages/sklearn/externals/joblib/parallel.py", line 28, in <module>
    from ._parallel_backends import (FallbackToBackend, MultiprocessingBackend,
  File "/opt/conda/lib/python3.8/site-packages/sklearn/externals/joblib/_parallel_backends.py", line 22, in <module>
    from .executor import get_memmapping_executor
  File "/opt/conda/lib/python3.8/site-packages/sklearn/externals/joblib/executor.py", line 14, in <module>
    from .externals.loky.reusable_executor import get_reusable_executor
  File "/opt/conda/lib/python3.8/site-packages/sklearn/externals/joblib/externals/loky/__init__.py", line 12, in <module>
    from .backend.reduction import set_loky_pickler
  File "/opt/conda/lib/python3.8/site-packages/sklearn/externals/joblib/externals/loky/backend/reduction.py", line 125, in <module>
    from sklearn.externals.joblib.externals import cloudpickle  # noqa: F401
  File "/opt/conda/lib/python3.8/site-packages/sklearn/externals/joblib/externals/cloudpickle/__init__.py", line 3, in <module>
    from .cloudpickle import *
  File "/opt/conda/lib/python3.8/site-packages/sklearn/externals/joblib/externals/cloudpickle/cloudpickle.py", line 167, in <module>
    _cell_set_template_code = _make_cell_set_template_code()
  File "/opt/conda/lib/python3.8/site-packages/sklearn/externals/joblib/externals/cloudpickle/cloudpickle.py", line 148, in _make_cell_set_template_code
    return types.CodeType(
TypeError: an integer is required (got type bytes)
```

It seems this is mainly a Python compatibility issue, so it's probably enough to update the XGBoost artefact in the test. These are saved in GCS though, so we'll take it from here to upload a fixed artefact there.

**Comment by adriangonz** [comment]
@Abhinavfreecodecamp sort of - it seems the XGBoost artefact used in the tests is not compatible with the changes you've made to the XGBoost server, so we'll need to update the artefact in GCS.

**Comment by Abhinavfreecodecamp** [comment]
@adriangonz got it, actually this is what I thought that's why I updated xgb version previously. Anyway I will list py3.8 compatible xgb below.

**Comment by adriangonz** [comment]
Hey @Abhinavfreecodecamp ,

You're right - all that was required was bumping the sklearn version on the xgboost runtime. I've also rebased from `master` to ensure it uses the latest version. 

I'll kick off again the integration tests :+1:

**Comment by Abhinavfreecodecamp** [comment]
Thanks @adriangonz

**Comment by seldondev** [comment]
@Abhinavfreecodecamp: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | e9be1a54d867d7ca684e809e072deebe38be5e57 | [link](http://dashboard-jx.34.121.152.94.nip.io/namespaces/jx/pipelineruns/seldon-core-pr-4910-integration-nvp8n) | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [jenkins-x/lighthouse](https://github.com/jenkins-x/lighthouse/issues/new?title=Command%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by Abhinavfreecodecamp** [comment]
@adriangonz still failing 🧐 strange as I recall sklearn 1.0 was compatible with py3.8

**Comment by adriangonz** [comment]
May still be a flaky one - let me have a look

**Comment by adriangonz** [comment]
Good news @Abhinavfreecodecamp - it seems like all tests are happy now! Thanks a lot for the time you've put into this one!

**Comment by Abhinavfreecodecamp** [comment]
@adriangonz thanks for helping me out.

**Pull request merged**